compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/inlineclasses/InlineClassAbi.kt

```diff
@@ -0,0 +1,121 @@
+import org.jetbrains.kotlin.load.java.JvmAbi
+import org.jetbrains.kotlin.name.Name
+
+/**
+ * Replace inline classes by their underlying types.
+ */
+fun IrType.unboxInlineClass() = InlineClassAbi.unboxType(this) ?: this
+
+object InlineClassAbi {
+    /**
+     * Unwraps inline class types to their underlying representation.
+     * Returns null if the type cannot be unboxed.
+     */
+    internal fun unboxType(type: IrType): IrType? {
+        val klass = type.classOrNull?.owner ?: return null
+        if (!klass.isInline) return null
+
+        val underlyingType = getUnderlyingType(klass).unboxInlineClass()
+        if (!type.isNullable())
+            return underlyingType
+        if (underlyingType.isNullable() || underlyingType.isPrimitiveType())
+            return null
+        return underlyingType.makeNullable()
+    }
+
+    /**
+     * Get the underlying type of an inline class based on the single argument to its
+     * primary constructor. This is what the current jvm backend does.
+     *
+     * Looking for a backing field does not work for unsigned types, which don't
+     * contain a field.
+     */
+    fun getUnderlyingType(irClass: IrClass): IrType {
+        require(irClass.isInline)
+        val primaryConstructor = irClass.constructors.single { it.isPrimary }
+        return primaryConstructor.valueParameters[0].type
+    }
+
+    /**
+     * Returns a mangled name for a function taking inline class arguments
+     * to avoid clashes between overloaded methods.
+     */
+    fun mangledNameFor(irFunction: IrFunction): Name {
+        val base = when {
+            irFunction is IrConstructor ->
+                "constructor"
+            irFunction.isGetter ->
+                JvmAbi.getterName(irFunction.propertyName.asString())
+            irFunction.isSetter ->
+                JvmAbi.setterName(irFunction.propertyName.asString())
+            irFunction.name.isSpecial ->
+                error("Unhandled special name in mangledNameFor: ${irFunction.name}")
+            else ->
+                irFunction.name.asString()
+        }
+
+        val suffix = when {
+            irFunction.fullValueParameterList.any { it.type.erasedUpperBound.isInline } ->
+                hashSuffix(irFunction)
+            (irFunction.parent as? IrClass)?.isInline == true -> "impl"
+            else -> return irFunction.name
+        }
+
+        return Name.identifier("$base-$suffix")
+    }
+
+    private val IrFunction.propertyName: Name
+        get() = (this as IrSimpleFunction).correspondingPropertySymbol!!.owner.name
+
+    private fun hashSuffix(irFunction: IrFunction) =
+        md5base64(irFunction.fullValueParameterList.joinToString { it.type.eraseToString() })
+
+    private fun IrType.eraseToString() = buildString {
+        append('L')
+        append(erasedUpperBound.fqNameWhenAvailable!!)
+        if (isNullable()) append('?')
+        append(';')
+    }
+}
+
+internal val IrFunction.hasInlineClassParameters: Boolean
```

Comment:
I propose to move all these properties which are used only in `InlineClassManager` closer to the usage and make them private

Reply:
Done